File: packages/ios/Sources/RichTextiOS/Parser.swift
import Foundation

public enum MarkdownParser {
    public static func parse(_ source: String, slackLineBreaks: Bool = true) -> DocumentNode {
        let normalized = source.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
        let lines = normalized.split(omittingEmptySubsequences: false, whereSeparator: \.isNewline)
            .map(String.init)
        var blocks: [BlockNode] = []
        var index = 0

        while index < lines.count {
            let line = lines[index]
            if line.trimmingCharacters(in: .whitespaces).isEmpty {
                index += 1
                continue
            }
            if let fence = matchFence(line) {
                let result = readCodeBlock(lines: lines, start: index, opening: fence)
                blocks.append(result.block)
                index = result.nextIndex
                continue
            }
            if let heading = matchHeading(line) {
                blocks.append(.heading(level: heading.level, children: parseInline(heading.content)))
                index += 1
                continue
            }
            if matchListItem(line) != nil {
                let result = readList(lines: lines, start: index, slackLineBreaks: slackLineBreaks)
                blocks.append(result.list)
                index = result.nextIndex
                continue
            }
            let result = readParagraph(lines: lines, start: index, slackLineBreaks: slackLineBreaks)
            blocks.append(result.paragraph)
            index = result.nextIndex
        }
        return DocumentNode(children: blocks)
    }

    public static func parseInline(_ source: String) -> [InlineNode] {
        parseInlineRange(source, start: source.startIndex, end: source.endIndex).nodes
    }
}

private let mentionRegex = try! NSRegularExpression(pattern: "^<@([^|>]+)\\|([^>]+)>")
private let headingRegex = try! NSRegularExpression(pattern: "^(#{1,6})[ \\t]+(.*)$")
private let unorderedRegex = try! NSRegularExpression(pattern: "^(\\s*)([-*+])[ \\t]+(.*)$")
private let orderedRegex = try! NSRegularExpression(pattern: "^(\\s*)(\\d+)\\.[ \\t]+(.*)$")
private let fenceRegex = try! NSRegularExpression(pattern: "^(`{3,}|~{3,})(.*)$")
private let escapable = Set(["\\", "`", "*", "_", "[", "]", "<", ">"])

private func match(_ regex: NSRegularExpression, in string: String) -> [String]? {
    let range = NSRange(string.startIndex..., in: string)
    guard let match = regex.firstMatch(in: string, range: range) else {
        return nil
    }
    var parts: [String] = []
    for index in 0..<match.numberOfRanges {
        guard let range = Range(match.range(at: index), in: string) else {
            parts.append("")
            continue
        }
        parts.append(String(string[range]))
    }
    return parts
}

private func matchFence(_ line: String) -> (marker: String, language: String?)? {
    guard let parts = match(fenceRegex, in: line) else {
        return nil
    }
    let language = parts.count > 2 ? parts[2].trimmingCharacters(in: .whitespaces) : ""
    return (parts[1], language.isEmpty ? nil : language)
}

private func matchHeading(_ line: String) -> (level: Int, content: String)? {
    guard let parts = match(headingRegex, in: line) else {
        return nil
    }
    return (min(parts[1].count, 6), parts[2])
}

private func matchListItem(_ line: String) -> (indent: String, ordered: Bool, content: String)? {
    if let parts = match(orderedRegex, in: line) {
        return (parts[1], true, parts[3])
    }
    if let parts = match(unorderedRegex, in: line) {
        return (parts[1], false, parts[3])
    }
    return nil
}

private func indentWidth(_ prefix: String) -> Int {
    prefix.reduce(0) { $0 + ($1 == "\t" ? 4 : 1) }
}

private func readCodeBlock(
    lines: [String],
    start: Int,
    opening: (marker: String, language: String?)
) -> (block: BlockNode, nextIndex: Int) {
    var body: [String] = []
    var index = start + 1
    while index < lines.count {
        let line = lines[index]
        if let closing = matchFence(line), closing.marker == opening.marker, closing.language == nil {
            index += 1
            break
        }
        body.append(line)
        index += 1
    }
    return (.codeBlock(language: opening.language, value: body.joined(separator: "\n")), index)
}

private func peekNonEmpty(lines: [String], from: Int) -> (index: Int, line: String)? {
    var index = from
    while index < lines.count {
        if !lines[index].trimmingCharacters(in: .whitespaces).isEmpty {
            return (index, lines[index])
        }
        index += 1
    }
    return nil
}

private func readParagraph(
    lines: [String],
    start: Int,
    slackLineBreaks: Bool
) -> (paragraph: BlockNode, nextIndex: Int) {
    var collected: [String] = []
    var index = start
    while index < lines.count {
        let line = lines[index]
        if line.trimmingCharacters(in: .whitespaces).isEmpty {
            break
        }
        if matchFence(line) != nil || matchHeading(line) != nil || matchListItem(line) != nil {
            break
        }
        collected.append(line)
        index += 1
    }
    let joined = slackLineBreaks ? collected.joined(separator: "\n") : collected.joined(separator: " ")
    return (.paragraph(MarkdownParser.parseInline(joined)), index)
}

private func readList(
    lines: [String],
    start: Int,
    slackLineBreaks: Bool
) -> (list: BlockNode, nextIndex: Int) {
    let first = lines[start]
    let firstMatch = matchListItem(first)!
    let ordered = firstMatch.ordered
    let baseIndent = indentWidth(firstMatch.indent)
    let startNumber: Int? = {
        if let parts = match(orderedRegex, in: first) {
            return Int(parts[2])
        }
        return nil
    }()
    var items: [ListItemNode] = []
    var index = start

    while index < lines.count {
        let line = lines[index]
        if line.trimmingCharacters(in: .whitespaces).isEmpty {
            guard let lookahead = peekNonEmpty(lines: lines, from: index + 1) else { break }
            guard let nextItem = matchListItem(lookahead.line),
                  indentWidth(nextItem.indent) >= baseIndent else { break }
            index += 1
            continue
        }
        guard let itemMatch = matchListItem(line) else { break }
        let indent = indentWidth(itemMatch.indent)
        if indent != baseIndent || itemMatch.ordered != ordered { break }
        let result = readListItem(lines: lines, start: index, itemIndent: indent, slackLineBreaks: slackLineBreaks)
        items.append(result.item)
        index = result.nextIndex
    }
    return (.list(ordered: ordered, start: ordered ? startNumber : nil, items: items), index)
}

private func readListItem(
    lines: [String],
    start: Int,
    itemIndent: Int,
    slackLineBreaks: Bool
) -> (item: ListItemNode, nextIndex: Int) {
    let first = matchListItem(lines[start])
    var paragraphLines = [first?.content ?? ""]
    var index = start + 1
    var children: [ListItemChild] = []

    func flushParagraph() {
        let text = paragraphLines.joined(separator: slackLineBreaks ? "\n" : " ")
            .replacingOccurrences(of: "\\s+$", with: "", options: .regularExpression)
        paragraphLines.removeAll()
        if !text.isEmpty {
            children.append(.paragraph(MarkdownParser.parseInline(text)))
        }
    }

    while index < lines.count {
        let line = lines[index]
        if line.trimmingCharacters(in: .whitespaces).isEmpty {
            guard let next = peekNonEmpty(lines: lines, from: index + 1) else { break }
            if let nextItem = matchListItem(next.line) {
                if indentWidth(nextItem.indent) <= itemIndent { break }
                flushParagraph()
                let nested = readList(lines: lines, start: next.index, slackLineBreaks: slackLineBreaks)
                if case let .list(ordered, startValue, items) = nested.list {
                    children.append(.list(ordered: ordered, start: startValue, items: items))
                }
                index = nested.nextIndex
                continue
            }
            index += 1
            continue
        }
        if let itemMatch = matchListItem(line) {
            let indent = indentWidth(itemMatch.indent)
            if indent == itemIndent { break }
            if indent > itemIndent {
                flushParagraph()
                let nested = readList(lines: lines, start: index, slackLineBreaks: slackLineBreaks)
                if case let .list(ordered, startValue, items) = nested.list {
                    children.append(.list(ordered: ordered, start: startValue, items: items))
                }
                index = nested.nextIndex
                continue
            }
            break
        }
        let continuation = line.prefix { $0 == " " || $0 == "\t" }
        if indentWidth(String(continuation)) > itemIndent {
            paragraphLines.append(line.trimmingCharacters(in: .whitespaces))
            index += 1
            continue
        }
        break
    }
    flushParagraph()
    if children.isEmpty {
        children.append(.paragraph([]))
    }
    return (ListItemNode(children: children), index)
}

private func parseInlineRange(
    _ source: String,
    start: String.Index,
    end: String.Index
) -> (nodes: [InlineNode], consumed: String.Index) {
    var nodes: [InlineNode] = []
    var index = start

    func pushText(_ value: String) {
        guard !value.isEmpty else { return }
        if case .text(let existing) = nodes.last {
            nodes[nodes.count - 1] = .text(existing + value)
        } else {
            nodes.append(.text(value))
        }
    }

    while index < end {
        let remaining = source[index..<end]
        if remaining.first == "\n" {
            nodes.append(.hardBreak)
            index = source.index(after: index)
            continue
        }
        if remaining.hasPrefix("\\"), remaining.count > 1 {
            let next = remaining[remaining.index(after: remaining.startIndex)]
            if escapable.contains(String(next)) {
                pushText(String(next))
                index = source.index(index, offsetBy: 2)
                continue
            }
        }
        if let mention = match(mentionRegex, in: String(remaining)) {
            nodes.append(.mention(MentionPayload(userId: mention[1], userName: mention[2])))
            index = source.index(index, offsetBy: mention[0].count)
            continue
        }
        if remaining.hasPrefix("`"), let code = readInlineCode(source, start: index, end: end) {
            nodes.append(.inlineCode(code.value))
            index = code.end
            continue
        }
        if remaining.hasPrefix("["), let link = readLink(source, start: index, end: end) {
            nodes.append(link.node)
            index = link.end
            continue
        }
        if remaining.hasPrefix("***") || remaining.hasPrefix("___") {
            let delimiter = String(remaining.prefix(3))
            if let closer = findCloser(source, start: source.index(index, offsetBy: 3), end: end, delimiter: delimiter) {
                let inner = parseInlineRange(source, start: source.index(index, offsetBy: 3), end: closer).nodes
                nodes.append(.bold([.italic(inner)]))
                index = source.index(closer, offsetBy: 3)
                continue
            }
        }
        if remaining.hasPrefix("**") || remaining.hasPrefix("__") {
            let delimiter = String(remaining.prefix(2))
            if let closer = findCloser(source, start: source.index(index, offsetBy: 2), end: end, delimiter: delimiter) {
                nodes.append(.bold(parseInlineRange(source, start: source.index(index, offsetBy: 2), end: closer).nodes))
                index = source.index(closer, offsetBy: 2)
                continue
            }
        }
        if remaining.hasPrefix("*") || remaining.hasPrefix("_") {
            let delimiter = String(remaining.first!)
            if !isFlankingOpener(source, index: index, delimiter: delimiter) {
                pushText(delimiter)
                index = source.index(after: index)
                continue
            }
            if let closer = findItalicCloser(
                source,
                start: source.index(after: index),
                end: end,
                delimiter: delimiter
            ) {
                nodes.append(.italic(parseInlineRange(source, start: source.index(after: index), end: closer).nodes))
                index = source.index(after: closer)
                continue
            }
        }
        pushText(String(remaining.first!))
        index = source.index(after: index)
    }
    return (nodes, index)
}

private func readInlineCode(
    _ source: String,
    start: String.Index,
    end: String.Index
) -> (value: String, end: String.Index)? {
    var cursor = start
    var tickCount = 0
    while cursor < end, source[cursor] == "`" {
        tickCount += 1
        cursor = source.index(after: cursor)
    }
    guard tickCount > 0 else { return nil }
    let opener = String(repeating: "`", count: tickCount)
    guard let closeRange = source.range(of: opener, range: cursor..<end) else { return nil }
    var value = String(source[cursor..<closeRange.lowerBound])
    if value.hasPrefix(" "), value.hasSuffix(" "), value.trimmingCharacters(in: .whitespaces).isEmpty == false {
        value = String(value.dropFirst().dropLast())
    }
    return (value, closeRange.upperBound)
}

private func readLink(
    _ source: String,
    start: String.Index,
    end: String.Index
) -> (node: InlineNode, end: String.Index)? {
    guard source[start] == "[",
          let labelEnd = findBalanced(source, start: source.index(after: start), end: end, open: "[", close: "]")
    else { return nil }
    let afterLabel = source.index(after: labelEnd)
    guard afterLabel < end, source[afterLabel] == "(",
          let destEnd = findBalanced(source, start: source.index(after: afterLabel), end: end, open: "(", close: ")")
    else { return nil }
    let label = String(source[source.index(after: start)..<labelEnd])
    let destination = String(source[source.index(after: afterLabel)..<destEnd])
        .trimmingCharacters(in: .whitespaces)
    let href = destination.split(whereSeparator: \.isWhitespace).first.map(String.init) ?? ""
    guard !href.isEmpty else { return nil }
    return (.link(href: href, children: MarkdownParser.parseInline(label)), source.index(after: destEnd))
}

private func findBalanced(
    _ source: String,
    start: String.Index,
    end: String.Index,
    open: Character,
    close: Character
) -> String.Index? {
    var depth = 1
    var index = start
    while index < end {
        if source[index] == "\\" {
            index = source.index(index, offsetBy: 2, limitedBy: end) ?? end
            continue
        }
        if source[index...].hasPrefix("<@") {
            if let mentionEnd = source[index..<end].firstIndex(of: ">") {
                index = source.index(after: mentionEnd)
                continue
            }
        }
        if source[index] == "`", let code = readInlineCode(source, start: index, end: end) {
            index = code.end
            continue
        }
        if source[index] == open {
            depth += 1
        } else if source[index] == close {
            depth -= 1
            if depth == 0 { return index }
        }
        index = source.index(after: index)
    }
    return nil
}

private func findCloser(
    _ source: String,
    start: String.Index,
    end: String.Index,
    delimiter: String
) -> String.Index? {
    var index = start
    while index < end {
        if source[index] == "\\" {
            index = source.index(index, offsetBy: 2, limitedBy: end) ?? end
            continue
        }
        if source[index...].hasPrefix("<@") {
            if let mentionEnd = source[index..<end].firstIndex(of: ">") {
                index = source.index(after: mentionEnd)
                continue
            }
        }
        if source[index] == "`", let code = readInlineCode(source, start: index, end: end) {
            index = code.end
            continue
        }
        if source[index...].hasPrefix(delimiter) {
            return index
        }
        index = source.index(after: index)
    }
    return nil
}

private func findItalicCloser(
    _ source: String,
    start: String.Index,
    end: String.Index,
    delimiter: String
) -> String.Index? {
    var index = start
    while index < end {
        if source[index] == "\\" {
            index = source.index(index, offsetBy: 2, limitedBy: end) ?? end
            continue
        }
        if source[index...].hasPrefix("<@") {
            if let mentionEnd = source[index..<end].firstIndex(of: ">") {
                index = source.index(after: mentionEnd)
                continue
            }
        }
        if source[index] == "`", let code = readInlineCode(source, start: index, end: end) {
            index = code.end
            continue
        }
        if source[index...].hasPrefix(delimiter + delimiter) {
            index = source.index(index, offsetBy: 2, limitedBy: end) ?? end
            continue
        }
        if String(source[index]) == delimiter {
            return index
        }
        index = source.index(after: index)
    }
    return nil
}

private func isFlankingOpener(_ source: String, index: String.Index, delimiter: String) -> Bool {
    let prev: Character = index == source.startIndex
        ? " "
        : source[source.index(before: index)]
    let nextIndex = source.index(after: index)
    let next: Character? = nextIndex < source.endIndex ? source[nextIndex] : nil
    guard let next, String(next) != delimiter, !next.isWhitespace else { return false }
    if delimiter == "_", prev.isLetter || prev.isNumber { return false }
    return true
}

File: packages/core/src/parser.ts
import type {
  BlockNode,
  BoldNode,
  CodeBlockNode,
  DocumentNode,
  HeadingNode,
  InlineNode,
  ItalicNode,
  ListItemNode,
  ListNode,
  ParagraphNode,
} from "./ast";
import { createDocument } from "./ast";

const MENTION_PATTERN = /^<@([^|>]+)\|([^>]+)>/;
const HEADING_PATTERN = /^(#{1,6})[ \t]+(.*)$/;
const UNORDERED_LIST_PATTERN = /^(\s*)([-*+])[ \t]+(.*)$/;
const ORDERED_LIST_PATTERN = /^(\s*)(\d+)\.[ \t]+(.*)$/;
const FENCE_PATTERN = /^(`{3,}|~{3,})(.*)$/;

const ESCAPABLE = new Set(["\\", "`", "*", "_", "[", "]", "<", ">"]);

export interface ParseOptions {
  /**
   * When true (default), a single newline inside a paragraph becomes a hard
   * break (Slack-style). When false, single newlines collapse to a space.
   */
  slackLineBreaks?: boolean;
}

export function parseMarkdown(
  source: string,
  options: ParseOptions = {},
): DocumentNode {
  const slackLineBreaks = options.slackLineBreaks ?? true;
  const normalized = source.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  const lines = normalized.split("\n");
  const blocks: BlockNode[] = [];
  let index = 0;

  while (index < lines.length) {
    const line = lines[index] ?? "";

    if (line.trim() === "") {
      index += 1;
      continue;
    }

    const fence = matchFence(line);
    if (fence) {
      const { block, nextIndex } = readCodeBlock(lines, index, fence);
      blocks.push(block);
      index = nextIndex;
      continue;
    }

    const heading = line.match(HEADING_PATTERN);
    if (heading) {
      const level = Math.min(heading[1]?.length ?? 1, 6) as HeadingNode["level"];
      const content = heading[2] ?? "";
      blocks.push({
        type: "heading",
        level,
        children: parseInline(content),
      });
      index += 1;
      continue;
    }

    const unordered = line.match(UNORDERED_LIST_PATTERN);
    const ordered = line.match(ORDERED_LIST_PATTERN);
    if (unordered || ordered) {
      const { list, nextIndex } = readList(lines, index, slackLineBreaks);
      blocks.push(list);
      index = nextIndex;
      continue;
    }

    const { paragraph, nextIndex } = readParagraph(lines, index, slackLineBreaks);
    blocks.push(paragraph);
    index = nextIndex;
  }

  return createDocument(blocks);
}

function matchFence(
  line: string,
): { marker: string; language: string | undefined } | undefined {
  const match = line.match(FENCE_PATTERN);
  if (!match) {
    return undefined;
  }
  const language = (match[2] ?? "").trim();
  return {
    marker: match[1] ?? "```",
    language: language.length > 0 ? language : undefined,
  };
}

function readCodeBlock(
  lines: string[],
  start: number,
  opening: { marker: string; language: string | undefined },
): { block: CodeBlockNode; nextIndex: number } {
  const body: string[] = [];
  let index = start + 1;
  while (index < lines.length) {
    const line = lines[index] ?? "";
    const closing = line.match(FENCE_PATTERN);
    if (closing && closing[1] === opening.marker && !(closing[2] ?? "").trim()) {
      index += 1;
      break;
    }
    body.push(line);
    index += 1;
  }
  return {
    block: {
      type: "codeBlock",
      language: opening.language || undefined,
      value: body.join("\n"),
    },
    nextIndex: index,
  };
}

function indentWidth(prefix: string): number {
  let width = 0;
  for (const character of prefix) {
    width += character === "\t" ? 4 : 1;
  }
  return width;
}

function readList(
  lines: string[],
  start: number,
  slackLineBreaks: boolean,
): { list: ListNode; nextIndex: number } {
  const first = lines[start] ?? "";
  const firstOrdered = first.match(ORDERED_LIST_PATTERN);
  const ordered = Boolean(firstOrdered);
  const baseIndent = indentWidth(
    (firstOrdered?.[1] ?? first.match(UNORDERED_LIST_PATTERN)?.[1] ?? ""),
  );
  const startNumber = firstOrdered ? Number.parseInt(firstOrdered[2] ?? "1", 10) : undefined;

  const items: ListItemNode[] = [];
  let index = start;

  while (index < lines.length) {
    const line = lines[index] ?? "";
    if (line.trim() === "") {
      const lookahead = peekNonEmpty(lines, index + 1);
      if (!lookahead) {
        break;
      }
      const nextItem = matchListItem(lookahead.line);
      if (!nextItem || indentWidth(nextItem.indent) < baseIndent) {
        break;
      }
      index += 1;
      continue;
    }

    const itemMatch = matchListItem(line);
    if (!itemMatch) {
      break;
    }
    const indent = indentWidth(itemMatch.indent);
    if (indent < baseIndent) {
      break;
    }
    if (indent > baseIndent) {
      break;
    }
    if (itemMatch.ordered !== ordered) {
      break;
    }

    const { item, nextIndex } = readListItem(lines, index, indent, slackLineBreaks);
    items.push(item);
    index = nextIndex;
  }

  return {
    list: {
      type: "list",
      ordered,
      start: ordered ? startNumber : undefined,
      children: items,
    },
    nextIndex: index,
  };
}

function matchListItem(
  line: string,
): { indent: string; ordered: boolean; content: string } | undefined {
  const ordered = line.match(ORDERED_LIST_PATTERN);
  if (ordered) {
    return { indent: ordered[1] ?? "", ordered: true, content: ordered[3] ?? "" };
  }
  const unordered = line.match(UNORDERED_LIST_PATTERN);
  if (unordered) {
    return { indent: unordered[1] ?? "", ordered: false, content: unordered[3] ?? "" };
  }
  return undefined;
}

function readListItem(
  lines: string[],
  start: number,
  itemIndent: number,
  slackLineBreaks: boolean,
): { item: ListItemNode; nextIndex: number } {
  const first = matchListItem(lines[start] ?? "");
  const paragraphLines: string[] = [first?.content ?? ""];
  let index = start + 1;
  const children: Array<ParagraphNode | ListNode> = [];

  while (index < lines.length) {
    const line = lines[index] ?? "";
    if (line.trim() === "") {
      const next = peekNonEmpty(lines, index + 1);
      if (!next) {
        break;
      }
      const nextItem = matchListItem(next.line);
      if (nextItem && indentWidth(nextItem.indent) <= itemIndent) {
        break;
      }
      if (nextItem && indentWidth(nextItem.indent) > itemIndent) {
        flushParagraph();
        const nested = readList(lines, next.index, slackLineBreaks);
        children.push(nested.list);
        index = nested.nextIndex;
        continue;
      }
      index += 1;
      continue;
    }

    const itemMatch = matchListItem(line);
    if (itemMatch && indentWidth(itemMatch.indent) === itemIndent) {
      break;
    }
    if (itemMatch && indentWidth(itemMatch.indent) > itemIndent) {
      flushParagraph();
      const nested = readList(lines, index, slackLineBreaks);
      children.push(nested.list);
      index = nested.nextIndex;
      continue;
    }
    if (itemMatch && indentWidth(itemMatch.indent) < itemIndent) {
      break;
    }

    const continuationIndent = line.match(/^(\s+)/)?.[1] ?? "";
    if (indentWidth(continuationIndent) > itemIndent) {
      paragraphLines.push(line.trim());
      index += 1;
      continue;
    }
    break;
  }

  flushParagraph();

  function flushParagraph(): void {
    const text = paragraphLines.join(slackLineBreaks ? "\n" : " ").trimEnd();
    paragraphLines.length = 0;
    if (text.length > 0) {
      children.push({
        type: "paragraph",
        children: parseInline(text),
      });
    }
  }

  if (children.length === 0) {
    children.push({ type: "paragraph", children: [] });
  }

  return {
    item: { type: "listItem", children },
    nextIndex: index,
  };
}

function peekNonEmpty(
  lines: string[],
  from: number,
): { index: number; line: string } | undefined {
  for (let index = from; index < lines.length; index += 1) {
    const line = lines[index] ?? "";
    if (line.trim() !== "") {
      return { index, line };
    }
  }
  return undefined;
}

function readParagraph(
  lines: string[],
  start: number,
  slackLineBreaks: boolean,
): { paragraph: ParagraphNode; nextIndex: number } {
  const collected: string[] = [];
  let index = start;
  while (index < lines.length) {
    const line = lines[index] ?? "";
    if (line.trim() === "") {
      break;
    }
    if (matchFence(line) || HEADING_PATTERN.test(line) || matchListItem(line)) {
      break;
    }
    collected.push(line);
    index += 1;
  }
  const joined = slackLineBreaks ? collected.join("\n") : collected.join(" ");
  return {
    paragraph: {
      type: "paragraph",
      children: parseInline(joined),
    },
    nextIndex: index,
  };
}

export function parseInline(source: string): InlineNode[] {
  return parseInlineRange(source, 0, source.length).nodes;
}

function parseInlineRange(
  source: string,
  start: number,
  end: number,
): { nodes: InlineNode[]; consumed: number } {
  const nodes: InlineNode[] = [];
  let index = start;

  const pushText = (value: string): void => {
    if (!value) {
      return;
    }
    const last = nodes[nodes.length - 1];
    if (last && last.type === "text") {
      last.value += value;
      return;
    }
    nodes.push({ type: "text", value });
  };

  while (index < end) {
    const remaining = source.slice(index, end);

    if (remaining.startsWith("\n")) {
      nodes.push({ type: "hardBreak" });
      index += 1;
      continue;
    }

    if (remaining.startsWith("\\") && remaining.length > 1) {
      const next = remaining[1] ?? "";
      if (ESCAPABLE.has(next)) {
        pushText(next);
        index += 2;
        continue;
      }
    }

    const mention = remaining.match(MENTION_PATTERN);
    if (mention) {
      nodes.push({
        type: "mention",
        userId: mention[1] ?? "",
        userName: mention[2] ?? "",
      });
      index += mention[0].length;
      continue;
    }

    if (remaining.startsWith("`")) {
      const code = readInlineCode(source, index, end);
      if (code) {
        nodes.push({ type: "inlineCode", value: code.value });
        index = code.end;
        continue;
      }
    }

    if (remaining.startsWith("[")) {
      const link = readLink(source, index, end);
      if (link) {
        nodes.push(link.node);
        index = link.end;
        continue;
      }
    }

    if (remaining.startsWith("***") || remaining.startsWith("___")) {
      const delimiter = remaining.slice(0, 3);
      const closer = findCloser(source, index + 3, end, delimiter);
      if (closer !== -1) {
        const inner = parseInlineRange(source, index + 3, closer).nodes;
        const italic: ItalicNode = { type: "italic", children: inner };
        const bold: BoldNode = { type: "bold", children: [italic] };
        nodes.push(bold);
        index = closer + 3;
        continue;
      }
    }

    if (remaining.startsWith("**") || remaining.startsWith("__")) {
      const delimiter = remaining.slice(0, 2);
      const closer = findCloser(source, index + 2, end, delimiter);
      if (closer !== -1) {
        nodes.push({
          type: "bold",
          children: parseInlineRange(source, index + 2, closer).nodes,
        });
        index = closer + 2;
        continue;
      }
    }

    if (remaining.startsWith("*") || remaining.startsWith("_")) {
      const delimiter = remaining[0] ?? "*";
      if (!isFlankingOpener(source, index, delimiter)) {
        pushText(delimiter);
        index += 1;
        continue;
      }
      const closer = findItalicCloser(source, index + 1, end, delimiter);
      if (closer !== -1) {
        nodes.push({
          type: "italic",
          children: parseInlineRange(source, index + 1, closer).nodes,
        });
        index = closer + 1;
        continue;
      }
    }

    pushText(remaining[0] ?? "");
    index += 1;
  }

  return { nodes, consumed: index - start };
}

function readInlineCode(
  source: string,
  start: number,
  end: number,
): { value: string; end: number } | undefined {
  let tickCount = 0;
  let cursor = start;
  while (cursor < end && source[cursor] === "`") {
    tickCount += 1;
    cursor += 1;
  }
  if (tickCount === 0) {
    return undefined;
  }
  const opener = "`".repeat(tickCount);
  const closeAt = source.indexOf(opener, cursor);
  if (closeAt === -1 || closeAt >= end) {
    return undefined;
  }
  let value = source.slice(cursor, closeAt);
  if (value.startsWith(" ") && value.endsWith(" ") && value.trim() !== "") {
    value = value.slice(1, -1);
  }
  return { value, end: closeAt + tickCount };
}

function readLink(
  source: string,
  start: number,
  end: number,
): { node: Extract<InlineNode, { type: "link" }>; end: number } | undefined {
  if (source[start] !== "[") {
    return undefined;
  }
  const labelEnd = findBalanced(source, start + 1, end, "[", "]");
  if (labelEnd === -1) {
    return undefined;
  }
  if (source[labelEnd + 1] !== "(") {
    return undefined;
  }
  const destEnd = findBalanced(source, labelEnd + 2, end, "(", ")");
  if (destEnd === -1) {
    return undefined;
  }
  const label = source.slice(start + 1, labelEnd);
  const destination = source.slice(labelEnd + 2, destEnd).trim();
  const href = destination.replace(/^<|>$/g, "").split(/\s+/)[0] ?? "";
  if (!href) {
    return undefined;
  }
  return {
    node: {
      type: "link",
      href,
      title: undefined,
      children: parseInline(label),
    },
    end: destEnd + 1,
  };
}

function findBalanced(
  source: string,
  start: number,
  end: number,
  open: string,
  close: string,
): number {
  let depth = 1;
  let index = start;
  while (index < end) {
    if (source[index] === "\\") {
      index += 2;
      continue;
    }
    if (source.startsWith("<@", index)) {
      const mentionEnd = source.indexOf(">", index);
      if (mentionEnd !== -1 && mentionEnd < end) {
        index = mentionEnd + 1;
        continue;
      }
    }
    if (source[index] === "`") {
      const code = readInlineCode(source, index, end);
      if (code) {
        index = code.end;
        continue;
      }
    }
    if (source[index] === open) {
      depth += 1;
    } else if (source[index] === close) {
      depth -= 1;
      if (depth === 0) {
        return index;
      }
    }
    index += 1;
  }
  return -1;
}

function findCloser(
  source: string,
  start: number,
  end: number,
  delimiter: string,
): number {
  let index = start;
  while (index <= end - delimiter.length) {
    if (source[index] === "\\") {
      index += 2;
      continue;
    }
    if (source.startsWith("<@", index)) {
      const mentionEnd = source.indexOf(">", index);
      if (mentionEnd !== -1 && mentionEnd < end) {
        index = mentionEnd + 1;
        continue;
      }
    }
    if (source[index] === "`") {
      const code = readInlineCode(source, index, end);
      if (code) {
        index = code.end;
        continue;
      }
    }
    if (source.startsWith(delimiter, index)) {
      return index;
    }
    index += 1;
  }
  return -1;
}

function findItalicCloser(
  source: string,
  start: number,
  end: number,
  delimiter: string,
): number {
  let index = start;
  while (index < end) {
    if (source[index] === "\\") {
      index += 2;
      continue;
    }
    if (source.startsWith("<@", index)) {
      const mentionEnd = source.indexOf(">", index);
      if (mentionEnd !== -1 && mentionEnd < end) {
        index = mentionEnd + 1;
        continue;
      }
    }
    if (source[index] === "`") {
      const code = readInlineCode(source, index, end);
      if (code) {
        index = code.end;
        continue;
      }
    }
    if (source.startsWith(delimiter + delimiter, index)) {
      index += 2;
      continue;
    }
    if (source[index] === delimiter) {
      return index;
    }
    index += 1;
  }
  return -1;
}

function isFlankingOpener(source: string, index: number, delimiter: string): boolean {
  const prev = index === 0 ? " " : source[index - 1] ?? " ";
  const next = source[index + 1] ?? "";
  if (!next || next === delimiter || /\s/.test(next)) {
    return false;
  }
  if (delimiter === "_" && /[A-Za-z0-9]/.test(prev)) {
    return false;
  }
  return true;
}
